src/exe/FreeC/Backend.hs

```diff
@@ -0,0 +1,125 @@
+module FreeC.Backend
+  ( Backend(..)
+  , backends
+  , coqBackend
+  , irBackend
+  )
+where
+
+import           Control.Monad.Extra            ( unlessM
+                                                , whenM
+                                                )
+import           Control.Monad.IO.Class
+import qualified Data.Map.Lazy                 as Map
+import           Data.Maybe                     ( isJust )
+import           System.Directory               ( createDirectoryIfMissing
+                                                , doesFileExist
+                                                , makeAbsolute
+                                                )
+import           System.FilePath
+
+import           FreeC.Application.Options
+import qualified FreeC.Backend.Coq.Base        as Coq.Base
+import qualified FreeC.Backend.Coq.Converter   as Coq.Converter
+import           FreeC.Backend.Coq.Pretty
+import qualified FreeC.IR.Syntax               as IR
+import           FreeC.Monad.Application
+import           FreeC.Pretty                   ( showPretty )
+
+-- | Data type that represents a backend.
+data Backend = Backend
+  { name          :: String
+    -- ^ The name of the backend.
+  , convertModule :: IR.Module -> Application String
+    -- ^ The conversion function that converts a module to program text.
+  , fileExtension :: String
+    -- ^ The file extension associated with the backend.
+  , specialAction :: Application ()
+    -- ^ An action that has to performed by the backend before conversion, e.g.
+    --   project file creation.
+  }
```

Comment:
The fields of the `Backend` and `Frontend` data types should have prefixes such as `backend` and `frontend`. Otherwise it wouldn't be possible to call a local variable `name` for example if the `FreeC.Backend` module is imported (since `-Wall -Werror` forbids shadowing).